### PR TITLE
Fix export of track-muted and set-track-muted.

### PR DESCRIPTION
**Pull request by jan-ivar**
Fixes https://github.com/w3c/mediacapture-main/issues/985

@marcoscaceres is this correct syntax?

***
<a href="https://pr-preview.s3.amazonaws.com/jan-ivar/mediacapture-main/pull/986.html" title="Last updated on Apr 22, 2024, 10:06 AM UTC (1294bec)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/mediacapture-main/986/6a4eccb...jan-ivar:1294bec.html" title="Last updated on Apr 22, 2024, 10:06 AM UTC (1294bec)">Diff</a>

### CODE REVIEW

**Comment by youennf** [comment]
We have other uses of `data-export` in the spec?
Should we update them as well?

**Comment by alvestrand** [comment]
Raising the @dontcallmedom signal for advice.

File: getusermedia.html [hunk 1]
```diff
@@ -897,7 +897,7 @@ <h4>Media Flow</h4>
         <p>There are two dimensions related to the media flow for a
         {{MediaStreamTrackState/"live"}} {{MediaStreamTrack}} : muted / not
         muted, and enabled / disabled.</p>
-        <p><dfn data-export id=
+        <p><dfn class="export" data-dfn-for="MediaStreamTrack" id=
```
Comment (jan-ivar) [inline getusermedia.html:900]:
```suggestion
        <p><dfn class="export" for="MediaStreamTrack" id=
```

File: getusermedia.html [hunk 2]
```diff
@@ -918,7 +918,8 @@ <h4>Media Flow</h4>
         <p>Whenever the [=User Agent=] initiates such a change, it MUST queue a
         task, using the user interaction task source, to [=set a track's muted
         state=] to the state desired by the user.</p>
-        <p>To <dfn class="abstract-op" id="set-track-muted">set a track's muted state</dfn> to
+        <p>To <dfn class="export" data-dfn-for="MediaStreamTrack"
```
Comment (jan-ivar) [inline getusermedia.html:921]:
```suggestion
        <p>To <dfn class="export" for="MediaStreamTrack"
```

Comment (youennf) [reply getusermedia.html]: 'dfn export for=MediaStreamTrack' works as per web codecs spec

**Comment by jan-ivar** [comment]
Hmm that didn't work either
> 1275.12-1275.142: error: Attribute “for” not allowed on element “dfn” at this point.
    "file:/home/runner/work/mediacapture-main/mediacapture-main.w3c/getusermedia.html":1294.15-1294.149: error: Attribute “for” not allowed on element “dfn” at this point.
  ❌  Not so good... please fix the issues above.

**Comment by dontcallmedom** [comment]
`for` is for bikeshed, respec needs `data-dfn-for`; I'll look into fixing this PR more generally early next week

File: getusermedia.html [hunk 3]
```diff
@@ -897,7 +897,7 @@ <h4>Media Flow</h4>
         <p>There are two dimensions related to the media flow for a
         {{MediaStreamTrackState/"live"}} {{MediaStreamTrack}} : muted / not
         muted, and enabled / disabled.</p>
-        <p><dfn data-export id=
+        <p><dfn class="export" for="MediaStreamTrack" id=
```
Comment (jan-ivar) [inline getusermedia.html:900]:
```suggestion
        <p><dfn export for=MediaStreamTrack id=
```

File: getusermedia.html [hunk 4]
```diff
@@ -918,7 +918,8 @@ <h4>Media Flow</h4>
         <p>Whenever the [=User Agent=] initiates such a change, it MUST queue a
         task, using the user interaction task source, to [=set a track's muted
         state=] to the state desired by the user.</p>
-        <p>To <dfn class="abstract-op" id="set-track-muted">set a track's muted state</dfn> to
+        <p>To <dfn class="export" for="MediaStreamTrack"
```
Comment (jan-ivar) [inline getusermedia.html:921]:
```suggestion
        <p>To <dfn export for=MediaStreamTrack
```

**Comment by jan-ivar** [comment]
@dontcallmedom ah thanks, will try that next

File: getusermedia.html [hunk 5]
```diff
@@ -897,7 +897,7 @@ <h4>Media Flow</h4>
         <p>There are two dimensions related to the media flow for a
         {{MediaStreamTrackState/"live"}} {{MediaStreamTrack}} : muted / not
         muted, and enabled / disabled.</p>
-        <p><dfn data-export id=
+        <p><dfn export for=MediaStreamTrack id=
```
Comment (jan-ivar) [inline getusermedia.html:900]:
```suggestion
        <p><dfn class="export" data-dfn-for="MediaStreamTrack" id=
```

**Comment by dontcallmedom** [comment]
looking at fixing the markup, there is an ambiguity in the language of the spec: sometimes "muted" is applied to the concept of source (e.g. "If source is stopped or muted, abort these steps"), sometimes to the concept of track (e.g. "A track sourced by a camera or microphone may be forcibly muted").

The `<dfn>` seems to allude (obliquely) to the source: "Muted refers to the input to the MediaStreamTrack".

I can propose a fix once this is clarified.

**Comment by alvestrand** [comment]
I *think* the idea was that "muted" refers to something that happens at the source, and is JS-visible through the "muted" attribute of the track (the source is ~never JS-visible). So "a track is muted" should probably be reworded as "the muted attribute is set on the track".
I think "input of the MediaStreamTrack" should generally be replaced with "source".

**Comment by jan-ivar** [comment]
>looking at fixing the markup, there is an ambiguity in the language of the spec: sometimes "muted" is applied to the concept of source (e.g. "If source is stopped or muted, abort these steps"), sometimes to the concept of track (e.g. "A track sourced by a camera or microphone may be forcibly muted").

I think that's largely an internal inconsistency, unrelated to exporting what to other specs is a track-related concept:
- concept https://w3c.github.io/mediacapture-main/#track-muted

>   Error: ROR] WebIDL identifier muted for MediaStreamTrack is defined multiple times

The error seems to come from the desire to add `data-dfn-for="MediaStreamTrack"` which makes it clash with MediaStreamTrack's `muted` WebIDL attribute of the same name:
- webidl https://w3c.github.io/mediacapture-main/#dom-mediastreamtrack-muted
  _"The muted attribute reflects whether the track is [muted](https://w3c.github.io/mediacapture-main/#track-muted). It MUST return [this](https://webidl.spec.whatwg.org/#this).[[[Muted]]](https://w3c.github.io/mediacapture-main/#dfn-muted)."_

Which reminds me there's also `this.[[Muted]]`:
- internal slot https://w3c.github.io/mediacapture-main/#dfn-muted

This same triad also exists for `enabled`:
- concept https://w3c.github.io/mediacapture-main/#track-enabled
- webidl https://w3c.github.io/mediacapture-main/#dom-mediastreamtrack-enabled
- internal slot https://w3c.github.io/mediacapture-main/#dfn-enabled

...which makes me question moving the muted concept to the source as a solution to the export problem.

How do other specs do this? E.g. HTML appears to have done away with internal slots? What's the state of the art here? 
- concept and internal slot https://html.spec.whatwg.org/multipage/media.html#concept-media-muted 
- webidl https://html.spec.whatwg.org/multipage/media.html#dom-media-muted

>  Error: ROR] Couldn't find "set a track's muted state" in this document or other cited documents: [dom], [html], [infra], [permissions], [permissions-policy], [webaudio], and [webidl].

The desire to add `data-dfn-for="MediaStreamTrack"` broke existing links in this spec (and other specs if we go ahead with it). Fixable by changing
```diff
- [=set a track's muted state=]
+ [=MediaStreamTrack/set a track's muted state=]
```
...in this spec and all other specs that reference it.

Choices here seem to be:
 1. Export the existing definitions without `data-dfn-for="MediaStreamTrack"` (is that frowned on?)
 2. add`data-dfn-for="MediaStreamTrack"` and rename the concept to "muted state" or something (ugh).
 3. Push back on mediaSession's need for the concept. E.g. could it say _"A user agent MAY expose microphone and camera state to web pages via the {{MediaStreamTrack/muted}} attribute."_?

**Comment by dontcallmedom** [comment]
exporting `muted` without a scope would be frowned upon, yes.

I've gone ahead and:
* made muted exported while distinct from the IDL attribute (using `data-dfn-type=dfn`)
* added a `<dfn>` for muted sources and used it in the couple of places where we had an autolink that didn't make sense in the context of MediaStreamTrack (which was the reason I raised this distinction in the first place)
* aligned the enabled and ended state to use the same pattern as muted

This will break autolinks in specs that were relying on these definitions

File: getusermedia.html [hunk 6]
```diff
@@ -845,15 +845,15 @@ <h4>Life-cycle</h4>
...
         <p>If all {{MediaStreamTrack}}s that are using the same
-        source are [= track/ended =], the source will be
+        source are [= MediaStreamTrack/ended =], the source will be
         [= source/stopped =].</p>
         <p>After the application has invoked the {{MediaStreamTrack/stop()}}
         method on a {{MediaStreamTrack}} object, or once the [=source=] of a
         {{MediaStreamTrack}} permanently ends production of live samples to its tracks,
         whichever is sooner, a {{MediaStreamTrack}} is said to be
-        <dfn id="track-ended" data-dfn-for="track" data-export>ended</dfn>.</p>
+        <dfn id="track-ended" data-dfn-for="MediaStreamTrack" data-dfn-type="dfn" data-export>ended</dfn>.</p>
```
Comment (marcoscaceres) [inline getusermedia.html:854]:
```suggestion
        <dfn data-dfn-for="MediaStreamTrack" class="export">ended</dfn>.</p>
```

File: getusermedia.html [hunk 7]
```diff
@@ -901,15 +901,15 @@ <h4>Media Flow</h4>
         <p>There are two dimensions related to the media flow for a
         {{MediaStreamTrackState/"live"}} {{MediaStreamTrack}} : muted / not
         muted, and enabled / disabled.</p>
-        <p><dfn data-export id=
+        <p><dfn class="export" data-dfn-for="MediaStreamTrack" data-dfn-type="dfn" id=
         "track-muted">Muted</dfn> refers to the input to the
```
Comment (marcoscaceres) [inline getusermedia.html:905]:
```suggestion
        <p><dfn class="export" data-dfn-for="MediaStreamTrack">Muted</dfn> refers to the input to the
```

File: getusermedia.html [hunk 8]
```diff
@@ -954,7 +955,7 @@ <h4>Media Flow</h4>
             <var>track</var>.</p>
           </li>
         </ol>
-        <p><dfn data-export id="track-enabled" data-lt="track enabled state|enabled" data-lt-noDefault>Enabled/disabled</dfn> on the other hand is
+        <p><dfn data-export id="track-enabled" data-dfn-for="MediaStreamTrack" data-dfn-type="dfn" data-lt="track enabled state|enabled" data-lt-noDefault>Enabled/disabled</dfn> on the other hand is
```
Comment (marcoscaceres) [inline getusermedia.html:958]:
```suggestion
        <p><dfn class="export" data-dfn-for="MediaStreamTrack" data-lt="track enabled state|enabled" data-lt-noDefault>Enabled/disabled</dfn> on the other hand is
```

Comment (marcoscaceres) [reply getusermedia.html:958]: @jan-ivar, the `data-lt="track enabled state|enabled"` doesn't make much sense. This should really be:

```HTML
<p><dfn class="export" data-dfn-for="MediaStreamTrack" data-lt="" data-lt-noDefault>track enabled state</dfn> can be either "enabled" or "disabled"...
```

See how HTML handles [Document/visibility state](https://html.spec.whatwg.org/multipage/interaction.html#visibility-state).

File: getusermedia.html [hunk 9]
```diff
@@ -299,14 +299,14 @@ <h2>{{MediaStream}}</h2>
       consumer.</p>
       <p>A {{MediaStream}} object is said to be <dfn data-dfn-for="stream" id=
       "stream-active">active</dfn> when it has at least one
-      {{MediaStreamTrack}} that has not [= track/ended =]. A {{MediaStream}} that does not
-      have any tracks or only has tracks that are [= track/ended =]
+      {{MediaStreamTrack}} that has not [= MediaStreamTrack/ended =]. A {{MediaStream}} that does not
```
Comment (marcoscaceres) [inline getusermedia.html:302]: Nit, global, no need for space after "[= ": 

```suggestion
      {{MediaStreamTrack}} that has not [=MediaStreamTrack/ended=]. A {{MediaStream}} that does not
```

File: getusermedia.html (hunk 6 above)
Comment (dontcallmedom) [reply getusermedia.html:854]: the id is used in other specs, I see no reason to break it; the `dfn-type` is needed to distinguish it from the IDL attribute of the name and scope
```suggestion
        <dfn id="track-ended" data-dfn-for="MediaStreamTrack" data-dfn-type="dfn" class=export>ended</dfn>.</p>
```

File: getusermedia.html (hunk 7 above)
Comment (dontcallmedom) [reply getusermedia.html:905]: likewise, `dfn-type` is needed to distinguish from the IDL attribute

**Comment by jan-ivar** [comment]
Thanks @dontcallmedom for the commits!

File: getusermedia.html [hunk 10]
```diff
@@ -299,14 +299,14 @@ <h2>{{MediaStream}}</h2>
...
+      {{MediaStreamTrack}} that has not [=MediaStreamTrack/ended=]. A {{MediaStream}} that does not
+      have any tracks or only has tracks that are [= MediaStreamTrack/ended =]
       is <dfn data-dfn-for="stream" id="stream-inactive">inactive</dfn>.</p>
       <p>A {{MediaStream}} object is said to be <dfn data-dfn-for=stream id=
       "stream-audible" data-dfn-for=stream>audible</dfn> when it has at least one
       {{MediaStreamTrack}} whose {{MediaStreamTrack/[[Kind]]}} is <a>"audio"</a>
-      that has not [= track/ended =]. A {{MediaStream}} that does not have any
-      audio tracks or only has audio tracks that are [= track/ended =] is
+      that has not [= MediaStreamTrack/ended =]. A {{MediaStream}} that does not have any
+      audio tracks or only has audio tracks that are [= MediaStreamTrack/ended =] is
```
Comment (marcoscaceres) [inline getusermedia.html:309]:
```suggestion
      that has not [=MediaStreamTrack/ended=]. A {{MediaStream}} that does not have any
      audio tracks or only has audio tracks that are [=MediaStreamTrack/ended=] is
```

File: getusermedia.html (hunk 6 above)
Comment (marcoscaceres) [reply getusermedia.html:854]: Hmm... is it defined twice for the two kinds?

File: getusermedia.html (hunk 8 above)
Comment (jan-ivar) [reply getusermedia.html:958]: This is unchanged by this PR, and I'm not sure what's the current respec way of doing things. @dontcallmedom can I bother you to address the feedback from @marcoscaceres here?

File: getusermedia.html (hunk 6 above)
Comment (dontcallmedom) [reply getusermedia.html:854]: once per kind, yes

File: getusermedia.html (hunk 8 above)
Comment (dontcallmedom) [reply getusermedia.html:958]:
```suggestion
        <p>The <dfn data-export id="track-enabled" data-dfn-for="MediaStreamTrack" data-dfn-type="dfn" >enabled</dfn> (resp. disabled) state on the other hand is
```

**Pull request merged**